### ISSUE TITLE
persistent timers with a large RandomizedDelaySec= are rarely run on systems that are rebooted frequently

### ISSUE THREAD

**Issue by mbiebl** [systemd/systemd#21166]
Hi,

we received a downstream bug report 
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=997943
where a user complained that the apt-daily.timer was not run on his system.

The timer looks like this
```
apt_2.3.11/debian/apt-daily.timer
[Timer]
OnCalendar=*-*-* 6,18:00
RandomizedDelaySec=12h
Persistent=true 
```

It has a large RandomizedDelaySec= to spread out the load over the whole day to avoid having millions of systems hitting the archive servers at the same time.

It turned out, that the user was frequently rebooting/shutting down his system and the systems themselves often aren't up fo a longer period of time.

With the changes from  #11608 , the `RandomizedDelaySec=12h` is applied on boot, resulting in the timer to rarely elapse.

This is a significant change of behaviour which affects several timers shipped in Debian that all use `RandomizedDelaySec=12h` 

So, while #11608 fixed an issue it also created a new one.

I wonder if systemd could be a bit more clever here and depending on how far  in the past the timestamp of the stampfile is, scale down `RandomizedDelaySec` accordingly. The further in the past, the shorter  `RandomizedDelaySec` would become so it would become more and more likely that the timer eventually fires.

On a related note: Afaics, we do not reapply `RandomizedDelaySec=` when a system resumes from suspend. This is  basically the same problem as #11608 tried to fix for the boot case.
Maybe we should also reapply a  `RandomizedDelaySec=` upon resume based on the time the system woke up?

**Comment by taro-yamada** [comment]
> depending on how far in the past the timestamp of the stampfile is, scale down `RandomizedDelaySec` accordingly.

It is very nice but I can't think of how to implement it...


> we do not reapply `RandomizedDelaySec=` when a system resumes from suspend.

Yes, we put that on hold at that point. [here](https://github.com/systemd/systemd/pull/11608#pullrequestreview-205845436)



I think the easiest way to mitigate this issue is adding an upper limit [here](https://github.com/systemd/systemd/blob/14efbfd96d67800e55137f89ec7a09d81049d176/src/core/timer.c#L425) like below.
```
usec_t elapse_after_next;
calendar_spec_next_usec(v->calendar_spec, v->next_elapse, &elapse_after_next);
if (v->next_elapse < rebased && rebased < elapse_after_next)
    v->next_elapse = rebased;
```
If your timer fired at 7:00 today
until 6:00 tomorrow, the rebase will be applied on boot so the timer will not fire immediately(same as #11608).
after 6:00 tomorrow, the rebase will not be applied on boot so the timer will fire immediately(same as before #11608).

**Comment by julian-klode** [comment]
I'd want timers to not fire immediately under any circumstances. This is highly problematic if you have a couple containers and they all run their timers during boot (or resume). Back when I suspended my laptop, I resumed in the morning and then the system was locking up while the timers on the host and a handful containers were catching up. Pain. I solved the issue by keeping it running 24/7 basically, wasting electricity :(

Also if you have scheduled down and uptimes (I don't know if anyone does), this can cause significant issues, for example, on mirror loads when the apt timers on Debian/Ubuntu systems run at the same time on all the machines you just booted and all start downloading from the server in parallel.

So basically I'd like to have the option to specify a `MinRandomizedDelaySec` too, that is always respected if the timer did not expire "naturally".

**Comment by me-and** [comment]
This has also caused me trouble: I have several systems I look after, which are frequently restarted, and I'd like jobs to run (say) every six months. I'd expected to be able to set `OnCalendar=semiannually`, `RandomizedDelaySec=6 months`, `Persistent=true` and `FixedRandomDelay=true` to ensure that each system will run their version of the job every six months, but across the entire set of systems, the jobs will be spread out over the specified period.

The only way I can get this to work currently is either to lose the ability to have common configuration for the timers, or to leave the systems running permanently.

It seems the issue is that fundamentally `RandomizedDelaySec` is doing two jobs at once: it's responsible for both (a) spreading out work to avoid overloading the local system, including immediately after startup, which it can currently achieve, and (b) spreading out work that would overload _external_ systems that serve multiple systemd systems (e.g. external databases, system administrators, &c.), which is broken if the systems restart too frequently relative to the configured delays.

Given the issue IMO is there being only a single config option for two different purposes, I'd propose a solution of adding a new configuration option to cover the second use case above, say `RandomizedOffsetSec`:

- `RandomizedDelaySec` retains its current meaning: after the next elapsing time has been determined – which might be immediate for a persistent timer that elapsed while the timer unit wasn't running – a random delay (subject to `FixedRandomDelay`) up to this value will be applied.
- `RandomizedOffsetSec` is applied before determining the next elapsing time: a random offset is calculated and added to the time given in `OnCalendar`.

I've excluded the non-calendar timers from `RandomizedOffsetSec` because I think – as with `Persistent` – it's less useful there, and the use case would be entirely covered by `RandomizedDelaySec`.

Handling the timer offsets would require a bit of care: for a semiannual job with a 6 month random offset, for example, if it's 2 January when the timer is enabled, and the random offset selected is 5 months, it should elapse on 1 May (1 Jan + 5 months) not 1 November (1 Jul + 5 months). I've also not yet thought about the best way to handle recalculating the offset; I expect either it should be always fixed in a similar way to `FixedRandomDelay` or there should be a new boolean toggle like `FixedRandomOffset`.

I think this would address the original issue, although it would need the affected timer units to switch from `RandomizedDelaySec` to `RandomizedOffsetSec`. It wouldn't break anything that's relying on the current behaviour, because the current behaviour isn't changed until/unless the new configuration options are used. And it would avoid @julian-klode's concerns about timers firing immediately; that would remain not being a problem using existing timer configuration.

If this seems at leas vaguely plausible as a high-level approach, I'm potentially willing (although I make no commitments) to have a go at writing the code and submitting a PR.

**Comment by anarcat** [comment]
i'm also suffering from that issue, as a blank normal laptop user, because i typically use my laptop only when i travel (either around the house or around the world, as opposed to my normal workstation). This means the laptop is typically off, and when it's on, it's on for a handful of hours at a time, which means it's quite likely i'm going to miss critical automated upgrades.

The solution suggested by @me-and is interesting, i wonder if `RandomizedOffsetSec` shouldn't be called `RandomizedJitterSec` instead, because `Offset` and `Delay` are kind of similar to me in naming. [Jitter](https://en.wikipedia.org/wiki/Jitter), however, has a specific meaning and the one we are looking for, I believe.

In the meantime, I guess the workaround for this is to reduce the `RandomizedDelaySec` parameter to whatever is the typical run time of your system... In my case, I think lowering down to 2-4h would be reasonable, for example.

**Comment by luiscastro193** [comment]
Adding `RandomizedJitterSec` would be nice in order to implement timers which spread execution during the day/week/month while also working on systems that are rebooted frequently.

I don't think there is a workaround in order to achieve both goals right now, since #3107 is also open. Or is it?

**Comment by AdrianVovk** [comment]
IMO this is a bug. The behavior we see in practice doesn't match what the documentation says will happen.

**Comment by anarcat** [comment]
we've been discussing this on matrix and i've stumbled upon this problem again in a different context (and totally forgot about this bug until @AdrianVovk pointed it out, amazingly). the long version is https://gitlab.torproject.org/tpo/tpa/team/-/issues/42043 but basically, i used RandomizedDelaySec to spread the load on a backup server across weeks/months, and it breaks down when the server reboots.

so, as @AdrianVovk says, i think there's a documentation problem here, at the very least, because I keep stumbling upon this, and can never recover, no matter how long i keep my eyes peeled on systemd.time(7) and systemd.timer(5). At the very least it should be made clearer that the RandomizedDelaySec applies after the last timer ran, or the last systemd start, which ever came last. Right now, there's no indication of this at all in the systemd.timer manual.

The FixedRandomDelay parameter has this note though:

>          For a given timer unit, the offset depends on
>           the machine ID, user identifier and timer name, which means
>           that it is stable between restarts of the manager.

... but I *really* don't feel this explains the situation clearly enough.

Now, that being said, I think a possible improvement here would be to have *some* way to do what I want (duh). I think I see why RandomizedDelaySec works the way it does: when you restart systemd, you don't want to immediately dispatch all those jobs, as they might *all* have expired and will overload everything, so perhaps keep the setting like this (but document that behavior properly) and instead add another OnCalendar setting that allow spreading an event across a given period. Another option that was discussed above is to ad a new setting like RandomizedDelaySec (i proposed RandomizedJitterSec in https://github.com/systemd/systemd/issues/21166#issuecomment-1196781473), but that this point I will let you fine artists decide the right way forward.

For now I've moved the randomization out of band, by picking an integer (for the "day of the month") and a string (for the "day of the week") in our configuration management, constant by machine but randomized across the fleet. It's not perfect, because it means we never have full backups running on the 29th, 30th or 31th (otherwise we might fail entirely on some months for hosts that pick those days), but it's good enough for our purposes (which is load spreading).

**Comment by AdrianVovk** [comment]
I think calling RandomizedOffsetSec "jitter" is wrong. Really RandomizedDelaySec is jitter.

For example, let's say you have OnCalendar=weekly. The true periodicity of the timer firing is once a week. A randomized offset doesn't change this, it just delays the starting point - you still fire once a week, but maybe you fire every Tuesday instead of every Monday. Meanwhile, RandomizedDelaySec changes the periodicity of the timer event: for a chosen random delay of N seconds, the timer fires "once a week, plus or minus N seconds". I'm pretty sure that's quite literally just the definition of jitter.

So, in a perfect world, RandomizedDelaySec gets renamed to RandomizedJitterSec, and then we can introduce a RandomizedOffsetSec like described above. But, this is quite messy - we'd need to introduce a bunch of aliases everywhere for backwards compat, and we'd need to rename "FixedRandomDelay" too. So, IMO calling them "delay" and  "offset" is probably good enough. People will need to read the docs to discover these settings anyways.

**Comment by anarcat** [comment]
the last thing i would do is block on bikeshedding the name of the thing, as long as there is a (documented) thing i'm happy. :)

**Comment by anarcat** [comment]
`FixedRandomDelay`, unexpectedly, does not work, and i thought i was pretty clear about that in https://github.com/systemd/systemd/issues/21166#issuecomment-2666831200 i did some tests with configs in https://github.com/anarcat/puppet-pgbackrest/commit/bc107d8aa4852d068e88aa57dbdbdc9306e7d7f9 and they don't survive systemd restarts.
